lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CTreeShapFeatureImportance.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace ml {
+namespace maths {
+
+using TRowItr = core::CDataFrame::TRowItr;
+
+std::pair<CTreeShapFeatureImportance::TDoubleVecVec, CTreeShapFeatureImportance::TDoubleVec>
+CTreeShapFeatureImportance::shap(const core::CDataFrame& frame,
+                                 const CDataFrameCategoryEncoder& encoder,
+                                 int numberFeatures) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode = this->samplesPerNode(tree, frame, encoder);
+        std::size_t maxDepth = this->updateNodeValues(tree, 0, samplesPerNode, 0);
+        maxDepthVec.push_back(maxDepth);
+        m_SamplesPerNode.emplace_back(std::move(samplesPerNode));
+    }
+
+    auto result = frame.readRows(
+        1, 0, frame.numberRows(),
```

Comment:
I think this should be thread-able no? Basically, you get a copy of state for each thread so as long as you can combine and this isn't mutating any other state you should be ok.

Reply:
Fixed.